Add extents tag

nilsvu opened the pull request:
## Proposed changes

Tag to store only the number of grid points in each dimension.
Useful for code that is agnostic to the Mesh and only needs to know
its extents. An example is the Schwarz linear solver that handles
overlaps with neighbors but doesn't need to know how those grid
points are laid out.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

geoffrey4444 commented on src/Domain/Tags.hpp:
```diff
@@ -107,6 +108,29 @@ struct Mesh : db::SimpleTag {
+/// \ingroup ComputationalDomainGroup
+/// \brief Retrieve the grid extents from the mesh
+template <size_t Dim>
+struct ExtentsFromMesh : Extents<Dim>, db::ComputeTag {
```
ExtentsComputeFromMesh or ExtentsFromMeshCompute?

nilsdeppe reviewed (changes requested):
Why can't you just grab the extents from the mesh? It seems like total overkill to have a compute tag just to retrieve a member variable.

nilsvu commented:
The Schwarz solver doesn't need to know how the grid points are laid out in the elements, it only works with the extents. So my intention with this tag is to avoid depending the Schwarz solver code on `Spectral` (though currently Domain depends on Spectral through Mesh, but that may be fixed in the future when we factor out libs from Domain to resolve cyclic includes). In fact the Schwarz solver test does not create a Mesh at all because it works with an explicit matrix and only knows its extents. So my intention is to decouple dependencies. What do you think?
The compute tag will only be added to the elliptic domain initializer, there's no need to add it to the evolution code.

nilsdeppe commented:
I think adding more compute tags is bad, especially silly ones like this that just add complexity to the code with no clear benefit.

geoffrey4444 commented:
What’s a situation where you would have extents that didn’t originally come from a mesh? Why not just pass the extents as a separate parameter if you want to factor in this way?

nilsvu committed:
Add extents tag

Tag to store only the number of grid points in each dimension.
Useful for code that is agnostic to the Mesh and only needs to know
its extents.

nilsvu commented:
@nilsdeppe @geoffrey4444 In response to your comments I removed the compute tag. I'll just depend the Schwarz code on the `Mesh`. I suppose I was overthinking the code design here; as @geoffrey4444 pointed out we won't realistically have extents without a mesh.

This PR is now reduced to adding the `domain::Tags::Extents` simple tag. I currently use it to store the number of overlap points of an element-centered subdomain with each neighbor. For this (and other data on overlaps) I have a `Schwarz::Tags::Overlaps<Tag>` prefix tag that stores a `FixedHashMap` indexed by overlap IDs, so I end up with a `Schwarz::Tags::Overlaps<domain::Tags::Extents<Dim>>` in the element's DataBox. Does that sound like a good way of structuring things to you?

nilsdeppe commented:
No, just add the tag locally that you need for the Schwarz smoother. All this use of prefix tags makes the code nearly impossible to understand. The boundary scheme code turned out to be so complicated that Will and I both gave up on understanding it and are just going to stop using it in the evolution code. Keep it simple, especially when there's a simple solution to the problem.

nilsvu commented:
ok I'll add a tag to the Schwarz solver then in another PR
